File: bridge/device_flow.py
from __future__ import annotations

from typing import Any
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit


def authorization_url(provider: str, verification_uri: str, user_code: str) -> str:
    if provider != "copilot" or not user_code:
        return verification_uri

    parts = urlsplit(verification_uri)
    query = parse_qsl(parts.query, keep_blank_values=True)
    query = [(key, value) for key, value in query if key != "user_code"]
    query.append(("user_code", user_code))
    return urlunsplit((parts.scheme, parts.netloc, parts.path, urlencode(query), parts.fragment))


def login_metadata(
    verification_uri: str,
    user_code: str,
    instructions: str,
) -> dict[str, Any]:
    metadata: dict[str, Any] = {
        "instructions": instructions,
        "verification_uri": verification_uri,
    }
    if user_code:
        metadata["user_code"] = user_code
    return metadata

File: internal/bridge/device_flow.go
package bridge

import (
	"net/url"
	"strings"
)

func authorizationURL(provider, verificationURI, userCode string) string {
	if provider != "copilot" || userCode == "" {
		return verificationURI
	}
	parsed, err := url.Parse(verificationURI)
	if err != nil {
		return verificationURI
	}
	query := parsed.Query()
	query.Set("user_code", userCode)
	parsed.RawQuery = query.Encode()
	return parsed.String()
}

func loginMetadata(verificationURI, userCode, instructions string) map[string]any {
	metadata := map[string]any{
		"instructions":     instructions,
		"verification_uri": verificationURI,
	}
	if strings.TrimSpace(userCode) != "" {
		metadata["user_code"] = userCode
	}
	return metadata
}
